mergify_engine/actions/queue.py

```diff
@@ -0,0 +1,109 @@
+# -*- encoding: utf-8 -*-
+#
+# Copyright © 2020 Mergify SAS
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+import dataclasses
+import typing
+
+import daiquiri
+import voluptuous
+
+from mergify_engine import check_api
+from mergify_engine import context
+from mergify_engine import queue
+from mergify_engine import rules
+from mergify_engine.actions import merge_base
+
+
+LOG = daiquiri.getLogger(__name__)
+
+
+class QueueAction(merge_base.MergeBaseAction):
+    validator = {
+        voluptuous.Required("name"): str,
+        voluptuous.Required(
+            "priority", default=merge_base.PriorityAliases.medium.value
+        ): voluptuous.All(
+            voluptuous.Any("low", "medium", "high", int),
+            voluptuous.Coerce(merge_base.Priority),
+            int,
+            voluptuous.Range(min=1, max=10000),
+        ),
+    }
+
+    queue_rule: rules.QueueRule = dataclasses.field(init=False)
+
+    def dispatch_config(self, mergify_config: typing.Dict):
+        self.config["bot_account"] = None
+        self.config["strict"] = "smart+ordered"
+        # NOTE(sileht): Don't allow rebase as it's too restricted
+        self.config["strict_method"] = "merge"
+
+        for rule in mergify_config["queue_rules"]:
+            if rule.name == self.config["name"]:
+                self.queue_rule = rule
+                self.config.update(rule.config)
+                break
+        else:
+            raise voluptuous.error.Invalid(f"{self.config['name']} queue not found")
+
+    @staticmethod
+    def _should_be_cancel(ctxt: context.Context, rule: rules.EvaluatedRule) -> bool:
+        # FIXME(sileht): !!!!!!!!!! IMPORTANT QUESTION !!!!!!!!!!!!
+        # Currently we remove a PR from the queue only if the pull_request_rules don't
+        # match. So if the pull request rules doesn't have check-success=foobar while the
+        # queue rule have it. When the check foobar fail, the PR is stuck in queue and
+        # unmergeable.
```

Comment:
We might need to have another action which is `dequeue` and describes when a PR should be removed from the queue.
Otherwise, it's too hackish to try to guess if e.g. you should wait or not for the CI to be retried or for a label to be added.